### ISSUE TITLE
Provide metalsmith.match (glob matching method)

### ISSUE THREAD

**Issue by webketje** [metalsmith/metalsmith#338]
With the plethora of file pattern matching options (regex, glob, mini/multi/nano-match, arrays or not), it would be interesting to add this method to the core.

**Benefits:**

- Canonical solution for matching file patterns with glob patterns (unifying plugin option API's)
- Plugins no longer need to re-implement their own solution and require it via NPM

**Drawbacks**
N/A

**Comment by ismay** [comment]
Personally I think I'd hold off on implementing core features until the more pressing matters have been resolved, and we have a clear long term vision of where we want to take the project.

For me the more pressing matters would be:

- Migrate our infrastructure, code and other areas away from segment
- Centralize all important metalsmith repositories in the metalsmith github and npm orgs
- Create a clear long term plan for the project

I think those points will have quite an effect on the project, and likely change our idea of what it needs. So any changes would ideally be made after we've taken those steps, in my opinion.

But @woodyrew has contributed a lot more to this repo, so I'm curious to hear what he thinks.

**Comment by webketje** [comment]
@ismay Yeah I just had to pin the idea down before it slipped, I share your line of thought.

**Comment by woodyrew** [comment]
As a long term concept, I agree it would provide a more homogeneous experience if plugins used a common glob matching facility. I think it would be prudent to use an already established package rather than re-invent the wheel.

With such things, I'm keen to keep the core repository as succinct and stable as possible so it would be worth reviewing pros and cons of the common packages that can provide this functionality. E.g. [Popularity](https://www.npmtrends.com/anymatch-vs-micromatch-vs-minimatch-vs-multimatch-vs-nanomatch), features, and their number of dependencies.

**Comment by ismay** [comment]
> As a long term concept, I agree it would provide a more homogeneous experience if plugins used a common glob matching facility. I think it would be prudent to use an already established package rather than re-invent the wheel.
> 
> With such things, I'm keen to keep the core repository as succinct and stable as possible so it would be worth reviewing pros and cons of the common packages that can provide this functionality. E.g. Popularity, features, and their number of dependencies.
> 

I agree, that would be my preferred approach as well.

**Comment by webketje** [comment]
Done preliminary research for the most important contenders: picomatch, nanomatch, micromatch, multimatch, minimatch.

TLDR: micromatch is the best compromise between perf, features & install size/ number of dependencies

Feature-wise, here is a good comparison: https://www.npmjs.com/package/picomatch#library-comparisons
I went ahead and created some perf tests for all contenders with common patterns in metalsmith: https://repl.it/@webketje/metalsmith-match-candidate?lite=true#test.js.

Take outs:

- minimatch does not support passing an array of patterns (nor thus, negated patterns)
- nanomatch & picomatch do not support brace expansion patterns (e.g. `**/*.{json,md}`)
- minimatch & multimatch are up to 10x slower than alternatives
- nanomatch is faster for simple expressions (no globstars etc), micromatch is faster for more complex globs.
- dependency & installed size: micromatch (61KB - 2 deps), nanomatch (86KB - 11 deps), multimatch (86 KB - 5 deps), picomatch (85.8KB - 0 deps)

**Comment by webketje** [comment]
Actually, there is another consideration:
The metalsmith prototype methods are either _build instructions_ (build, process) or _build option get/setters_ (use, metadata, source, concurrency, destination),  with the notable exception of _.path_, which is a helper.
The match method is a helper too. So *domain-wise* it would be more appropriate to provide it either:

- A) as `Metalsmith.match` ('static' method)
- B) as `Metalsmith.prototype.helpers.match`

The problem with A) is that the Metalsmith constructor func is not accessible from the plugin signature (so plugin authors would need to `require('metalsmith')` which is IMO a no-go: having metalsmith in peerDependencies is better), while B) introduces an extra namespace.

The current proposal would allow plugin authors to do:

```js
function plugin(files, metalsmith) {
  metalsmith
     .match('**/*.md')
     .forEach(path => files[path] /* do something */)
}
```
while the B) option would be a bit more verbose 
```js
function plugin(files, metalsmith) {
  const { helper } = metalsmith
  helper
     .match('**/*.md')
     .forEach(path => files[path] /* do something */)
}
```
The advantage of B) is that it provides a path for adding more helpers, but.. do we really need that?

**Comment by wernerglinka** [comment]
I don't have a preference but I completely agree with your simplicity and declaratively argument.

Werner Glinka
***@***.***
650 867 0778
